pipeline/eval/metrics.py

```diff
@@ -0,0 +1,694 @@
+import json
+import os
+import statistics
+from collections import defaultdict
+from dataclasses import dataclass
+import time
+from pathlib import Path
+from statistics import mean
+from typing import Any, List, Iterable
+
+import sacrebleu.metrics.base
+import toolz
+from sacrebleu.metrics.bleu import BLEU
+from sacrebleu.metrics.chrf import CHRF
+
+from pipeline.common.logging import get_logger
+from pipeline.eval.langs import COMET22_SUPPORT, METRICX24_SUPPORT
+
+logger = get_logger(__file__)
+
+
+@dataclass
+class MetricResults:
+    name: str
+    segment_scores: list[Any]
+    corpus_score: float
+    details: dict[str, Any]
+
+
+class Metric:
+    name = None
+
+    @staticmethod
+    def supports_lang(src_lang: str, trg_lang: str) -> bool:
+        ...
+
+
+class RegularMetric(Metric):
+    def score(
+        self,
+        src_lang: str,
+        trg_lang: str,
+        source_texts: list[str],
+        translated_texts: list[str],
+        reference_texts: list[str],
+    ) -> MetricResults:
+        ...
+
+
+class ReferencelessMetric(Metric):
+    def score_qe(
+        self, src_lang: str, trg_lang: str, source_texts: list[str], translated_texts: list[str]
+    ) -> MetricResults:
+        ...
+
+
+class SacrebleuMetric(RegularMetric):
+    name = None
+
+    def __init__(self):
+        self.metric: sacrebleu.metrics.base.Metric = None
+
+    @staticmethod
+    def supports_lang(src_lang: str, trg_lang: str) -> bool:
```

Comment:
A slight nuance here is that, even if BLEU as metric supports every language, the tokenization does not. Instead of maybe trying to list all supported languages by the default tokenization, should we instead switch to a SentencePiece based tokenization? sacreBLEU recently merged spBLEU-1K for 1K languages but it's still not released, probably `flores200` is a better choice for now.